web/app/Http/Controllers/PageController.php

```diff
@@ -4,19 +4,205 @@
+            $page = Page::findSlug($site_id, $path, $wikitext, $html);
+        } elseif ($path_type === 'id') {
+            $page = Page::findId($site_id, (int) $path, $wikitext, $html);
+        } else {
+            return apierror(400, APIError::INVALID_PAGE_PATH);
+        }
+
+        if ($page === null) {
+            return apierror(404, APIError::PAGE_NOT_FOUND);
+        }
+
+        return $page;
+    }
+
+    /**
+     * Gets a page.
+     * Endpoint: `GET:/page/{path_type}/{path}` | `pageGet`
+     */
+    public function pageGet(Request $request, string $path_type, string $path)
+    {
+        $output_type = $this->resolvePageType($request);
+        if ($output_type === null) {
+            return apierror(400, APIError::INVALID_PAGE_TYPE);
+        }
+
+        $page = $this->resolvePage(
+            $path_type,
+            $path,
+            $output_type['wikitext'],
+            $output_type['html'],
+        );
+
+        // api error was returned
+        if ($page instanceof Response) {
+            return $page;
+        }
+
+        $output = [];
+
+        if ($output_type['metadata']) {
+            $avatars = (bool) $request->query('avatars', false);
+
+            // TODO: get creator through earliest revision
+            // (unless we implement the authors field first)
+
+            $updater = User::findId($page->revision_user_id);
+
+            if ($updater !== null) {
+                $updater = $updater->toApiArray($avatars);
+            }
+
+            $output = array_merge($output, [
+                'id' => $page->id,
+                'slug' => $page->slug,
+                'category' => $page->page_category_id,
+                'parent' => null, // TODO
+                'children' => [], // TODO
+                'title' => $page->title,
+                'altTitle' => $page->alt_title,
+                'tags' => $page->tags,
+                'score' => 0, // TODO
+                'created' => $page->page_created_at,
+                'creator' => null, // TODO
+                'updated' => $page->lastUpdated(),
+                'updater' => $updater,
+            ]);
+        }
+
+        if ($output_type['wikitext']) {
+            $output = array_merge($output, [
+                'wikitext' => $page->wikitext,
+            ]);
+        }
+
+        if ($output_type['html']) {
+            $output = array_merge($output, [
+                'html' => $page->compiled_html,
+            ]);
+        }
+
+        return new Response($output, 200);
+    }
+
     /** Returns a `view` for the current page. */
-    public function show(): View
+    public function show(?string $path = null): View
     {
-        $values = LegacyTools::generateScreenVars();
+        // TODO: description, image, twitter, etc.
+        // TODO: site theming
+        // TODO: favicons
+        // TODO: header image/text + subtitle management
+        // TODO: navbar items
+        // TODO: private sites
```

Comment:
Oof not really looking forward to this. It's not used a lot and will involve us really getting into the details of permissions and starting implementations there.

That said, I do look forward to actually making a cohesive permission system that is extensible and isn't brittle.